### PR TITLE
Adds pagination to currency listing GET API

### PR DESCRIPTION
**Pull request by ghaleroshan**
- [x] Limit of 10 rates per page

### CODE REVIEW

File: src/exchangeRates.js [hunk 1]
```diff
@@ -60,7 +60,7 @@ async function get(params) {
 
 async function getMultiple() {
   let allExchangeRates = await db.query(
-    `SELECT from_currency, to_currency, rate, created_at FROM exchange_rates`,
+    `SELECT from_currency, to_currency, rate, created_at FROM exchange_rates LIMIT 10`,
```
Comment (geshan) [inline src/exchangeRates.js]: this is not going to paginate, add page in the URL and an offset here too. then add tests for the same, please!

File: src/exchangeRates.js [hunk 2]
```diff
@@ -1,19 +1,23 @@
+
+
```
Comment (geshan) [inline src/exchangeRates.js]: please remove these blank lines

File: src/exchangeRates.js [hunk 3]
```diff
@@ -1,19 +1,23 @@
...
 const axios = require('axios');
 const config = require('./config');
 const db = require('namshi-node-mysql')(config.db);
 const _ = require('lodash');
 const {httpError} = require('expressjs-utils');
 
+
+
```
Comment (geshan) [inline src/exchangeRates.js]: please remove these too :)

File: src/exchangeRates.js [hunk 4]
```diff
@@ -45,31 +49,38 @@ async function get(params) {
...
     return {fromCurrency, toCurrency, onDate, rate};
   }
-  
+
   return getExternal(fromCurrency, toCurrency, onDate);
 }
 
-async function getMultiple() {
+async function getMultiple(currentPage) {
+  let itemsPerPage = 10;
```
Comment (geshan) [inline src/exchangeRates.js]: can you please put this in config.js and pull it from there.

File: src/exchangeRates.js [hunk 5]
```diff
@@ -45,31 +49,38 @@ async function get(params) {
...
+
+  let offset = (currentPage - 1) * itemsPerPage;
+
+  console.log(offset);
+
   let allExchangeRates = await db.query(
-    `SELECT from_currency, to_currency, rate, created_at FROM exchange_rates`,
-  );
-  if (allExchangeRates.length) {
+      `SELECT from_currency, to_currency, rate, created_at FROM exchange_rates ORDER BY rate desc LIMIT ${offset}, ${itemsPerPage}`);
```
Comment (geshan) [inline src/exchangeRates.js]: order by will not be needed.

File: test/exchnageRatesTest.js [hunk 6]
```diff
@@ -101,16 +101,31 @@ describe('exchangeRates', () => {
...
         assert.equal(err.statusCode, 400);
-        assert.equal(err.message, 'Error in fetching rate');        
+        assert.equal(err.message, 'Error in fetching rate');
       }
     });
+
+      it('should paginate the result with the limit of 10 rows per page', async () => {
+        try {
+          mysqlStub.query = (query, params) => {
+            if (query.startsWith(`SELECT from_currency, to_currency`)){
```
Comment (geshan) [inline test/exchnageRatesTest.js]: testing how the query ends with would be needed here.

File: index.js [hunk 7]
```diff
@@ -14,16 +14,17 @@ app.get('/', (req,res) => {
 
 app.get('/api/convert/:fromCurrency/:toCurrency/:onDate', async (req, res) => {
   console.log(`Api hit`, req.params);
-  
   res.json(await exchangeRates.get(req.params));
 });
 
+
 app.get('/api/rates', async (req,res) => {
   //res.json({message: `All Caches Rates will come here`});
```
Comment (geshan) [inline index.js]: can you remove this comment too :)

File: index.js [hunk 8]
```diff
@@ -14,16 +14,17 @@ app.get('/', (req,res) => {
...
-  res.json(await exchangeRates.getMultiple());
+  res.json(await exchangeRates.getMultiple(req.query.page || 1));
 });
-
 expressUtils.hc(app);
 expressUtils.static(app);
 expressUtils.errorHandler(app);
 expressUtils.start(app, 8080, 'dev');
+
+
```
Comment (geshan) [inline index.js]: needless empty lines.

File: src/exchangeRates.js (hunk 4 above)
Comment (ghaleroshan) [reply src/exchangeRates.js]: Do you mean for itemsPerPage only?

Comment (geshan) [reply src/exchangeRates.js]: Yes

File: index.js [hunk 9]
```diff
@@ -14,16 +14,15 @@ app.get('/', (req,res) => {
...
   //res.json({message: `All Caches Rates will come here`});
-  res.json(await exchangeRates.getMultiple());
+  res.json(await exchangeRates.getMultiple(req.query.page || 1));
 });
 
 expressUtils.hc(app);
 expressUtils.static(app);
 expressUtils.errorHandler(app);
-expressUtils.start(app, 8080, 'dev');
+expressUtils.start(app, 8080, 'dev');
```
Comment (geshan) [inline index.js]: Don't delete last empty line

File: src/config.js [hunk 10]
```diff
@@ -9,8 +9,10 @@ const config = {
   currencyConverterApi: {
     baseUrl: 'https://free.currconv.com/api/v7',
     key: 'ca0133cec785f144a8c5'
-  }
-}
+  },
+  itemsPerPage: 10
+};
+
```
Comment (geshan) [inline src/config.js]: Line 15 can be removed :)

File: src/exchangeRates.js [hunk 11]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (geshan) [inline src/exchangeRates.js]: Line 1 can be removed

File: src/exchangeRates.js [hunk 12]
```diff
@@ -45,31 +46,33 @@ async function get(params) {
...
   }
-  
   return [];
 }
 
 module.exports = {
   get,
   getMultiple,
-}
+};
```
Comment (geshan) [inline src/exchangeRates.js]: Need empty line at end

File: test/exchnageRatesTest.js [hunk 13]
```diff
@@ -12,15 +12,15 @@ const today = new Date().toISOString().split('T')[0];
 
 describe('exchangeRates', () => {
   describe('get', () => {
-    it('should use given params and if no results are in db, inserts new rate to db while returing it', async () => {
+    it('should use given params and if no results are in db, inserts new rate to db while returning it', async () => {
```
Comment (geshan) [inline test/exchnageRatesTest.js]: Thanks!

File: test/exchnageRatesTest.js [hunk 14]
```diff
@@ -101,16 +101,32 @@ describe('exchangeRates', () => {
...
-        assert.equal(err.message, 'Error in fetching rate');        
+        assert.equal(err.message, 'Error in fetching rate');
       }
     });
+
+      it('should paginate the result with the limit of 10 rows per page', async () => {
+        try {
+          mysqlStub.query = (query, params) => {
+            if (query.startsWith(`SELECT from_currency, to_currency`)
+                && query.endsWith(`LIMIT ${exchangeRates.offset}, ${exchangeRates.itemsPerPage}`)){
```
Comment (geshan) [inline test/exchnageRatesTest.js]: Use exact numbers like 0,10 not vars.

File: test/exchnageRatesTest.js [hunk 15]
```diff
@@ -112,14 +112,12 @@ describe('exchangeRates', () => {
...
     });
-
       it('should paginate the result with the limit of 10 rows per page', async () => {
         try {
           mysqlStub.query = (query, params) => {
-            if (query.startsWith(`SELECT from_currency, to_currency`)
-                && query.endsWith(`LIMIT ${exchangeRates.offset}, ${exchangeRates.itemsPerPage}`)){
-              return[];
-            }
+            if (query.startsWith(`SELECT from_currency, to_currency`
```
Comment (geshan) [inline test/exchnageRatesTest.js]: Missing {} for if

File: .gitignore [hunk 16]
```diff
@@ -1,3 +1,4 @@
+.idea
```
Comment (geshan) [inline .gitignore]: Better add it in your global gitignore.

File: src/config.js [hunk 17]
```diff
@@ -9,8 +9,8 @@ const config = {
   currencyConverterApi: {
     baseUrl: 'https://free.currconv.com/api/v7',
     key: 'ca0133cec785f144a8c5'
-  }
-}
+  },
+  itemsPerPage: 10
```
Comment (geshan) [inline src/config.js]: Can you please make it configurable from an env variable too, similar to db creds above?

File: src/exchangeRates.js [hunk 18]
```diff
@@ -45,31 +45,34 @@ async function get(params) {
...
-async function getMultiple() {
+async function getMultiple(currentPage) {
+
+  let offset = (currentPage - 1) * [config.itemsPerPage];
+
   let allExchangeRates = await db.query(
-    `SELECT from_currency, to_currency, rate, created_at FROM exchange_rates`,
-  );
-  if (allExchangeRates.length) {
+      `SELECT from_currency, to_currency, rate, created_at FROM exchange_rates LIMIT ${offset}, ${config.itemsPerPage}`);
```
Comment (geshan) [inline src/exchangeRates.js]: Can you please select on_date field and use a parameterized query for this with `?` for offset and limit.

File: test/exchnageRatesTest.js [hunk 19]
```diff
@@ -101,16 +101,31 @@ describe('exchangeRates', () => {
...
         assert.equal(err.statusCode, 400);
-        assert.equal(err.message, 'Error in fetching rate');        
+        assert.equal(err.message, 'Error in fetching rate');
       }
     });
+      it('should paginate the result with the limit of 10 rows per page', async () => {
+        try {
+          mysqlStub.query = (query, params) => {
+            if (query.startsWith(`SELECT from_currency, to_currency`
+                && query.endsWith(`LIMIT 0 10`)))
```
Comment (geshan) [inline test/exchnageRatesTest.js]: The bracket seems to be off, is there a syntax error here?

**Pull request merged**